### ISSUE TITLE
Using `AutomaticKeepAliveClientMixin` on `SliverList` item throws "RenderBox was not laid out: RenderParagraph#d50c4 NEEDS-LAYOUT NEEDS-PAINT"

### ISSUE THREAD

**Issue by code1up** [flutter/flutter#21601]
I have a `SliverList` that has items built using `FutureBuilder`. The items depend on a web API. I have replicated the problem in the `main.dart`:

```dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

void main() => runApp(Randoms());

class Randoms extends StatelessWidget {
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Randoms',
      theme: ThemeData(
        primarySwatch: Colors.orange,
      ),
      home: Page(title: 'Randoms'),
    );
  }
}

class Page extends StatefulWidget {
  Page({
    Key key,
    this.title,
  }) : super(key: key);

  final String title;

  _PageState createState() => _PageState();
}

class _PageState extends State<Page> {
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: CustomScrollView(
        slivers: [
          _buildList(),
        ],
      ),
    );
  }

  SliverList _buildList() {
    return SliverList(
      delegate: SliverChildBuilderDelegate(
        _buildRow,
        childCount: 50,
      ),
    );
  }

  Widget _buildRow(buildContext, index) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Random(
        index: index,
      ),
    );
  }
}

class Random extends StatefulWidget {
  final int index;

  Random({
    this.index,
  });

  _RandomState createState() => _RandomState();
}

class _RandomState extends State<Random> {
  Widget build(BuildContext context) {
    final url =
        'https://randomuser.me/api?page=${widget.index + 1}&results=1&seed=seeded';

    return FutureBuilder<RandomPerson>(
      future: _fetchRandomPerson(url),
      builder: _buildView,
    );
  }

  Future<RandomPerson> _fetchRandomPerson(String url) async {
    final response = await http.get(url);
    final body = json.decode(response.body);
    final results = body['results'];

    return RandomPerson.fromJson(results[0]);
  }

  Widget _buildView(
    BuildContext context,
    AsyncSnapshot<RandomPerson> snapshot,
  ) {
    final key = Key('${widget.index + 1}');

    return RandomRow(
      key: key,
      index: widget.index,
      snapshot: snapshot,
    );
  }
}

class RandomRow extends StatefulWidget {
  final AsyncSnapshot<RandomPerson> snapshot;
  final int index;

  RandomRow({
    Key key,
    this.index,
    this.snapshot,
  }) : super(key: key);

  RandomRowState createState() => RandomRowState();
}

class RandomRowState extends State<RandomRow>
    with AutomaticKeepAliveClientMixin {
  void initState() {
    super.initState();

    print('initState ${widget.index + 1}');
  }

  Widget build(BuildContext context) {
    super.build(context);

    final theme = Theme.of(context);

    if (widget.snapshot.hasError) {
      return Center(
        child: Text(
          widget.snapshot.error.toString(),
          style: theme.textTheme.display1,
        ),
      );
    }

    if (!widget.snapshot.hasData) {
      return Center(
        child: Text(
          'WAITING',
          style: theme.textTheme.display1,
        ),
      );
    }

    final person = widget.snapshot.data;

    return Text(
      '${widget.index + 1} ${person.firstName}',
      style: theme.textTheme.display1,
    );
  }

  bool get wantKeepAlive => true;
}

class RandomPerson {
  final String firstName;
  final String lastName;

  RandomPerson({
    this.firstName,
    this.lastName,
  });

  factory RandomPerson.fromJson(Map<String, dynamic> json) {
    final name = json['name'];

    return RandomPerson(
      firstName: name['first'],
      lastName: name['last'],
    );
  }
}
```

Below is the error I see in the VS Code debug window (and why I'm here). I've tried lots of different approaches to make this work. The app itself is functional but having the underlying framework throw exceptions must mean something is wrong.

You can reproduce the error by scrolling up and down. My **guess** is that the future is being returned and when the item is no longer visible but being 'kept alive' and therefore cannot be drawn. That's a guess!

Any help appreciated. This feels like an common requirement - so must be a problem of my own making. Or can these exceptions be ignored (seems unlikely)?

Thanks.
```
flutter: ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following assertion was thrown during a scheduler callback:
flutter: RenderBox was not laid out: RenderParagraph#d50c4 NEEDS-LAYOUT NEEDS-PAINT
flutter: 'package:flutter/src/rendering/box.dart': Failed assertion: line 1492 pos 12: 'hasSize'
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      RenderBox.size (package:flutter/src/rendering/box.dart:1492:12)
flutter: #3      RenderBox.semanticBounds (package:flutter/src/rendering/box.dart:1620:44)
flutter: #4      _SemanticsGeometry._computeValues (package:flutter/src/rendering/object.dart:3394:100)
flutter: #5      new _SemanticsGeometry (package:flutter/src/rendering/object.dart:3352:5)
flutter: #6      _SwitchableSemanticsFragment.compileChildren (package:flutter/src/rendering/object.dart:3265:15)
flutter: #7      _SyncIterator.moveNext (dart:core/runtime/libcore_patch.dart:151:12)
flutter: #8      _SyncIterator.moveNext (dart:core/runtime/libcore_patch.dart:144:32)
flutter: #9      List.addAll (dart:core/runtime/libgrowable_array.dart:195:17)
flutter: #10     _SwitchableSemanticsFragment.compileChildren (package:flutter/src/rendering/object.dart:3291:16)
flutter: #11     _SyncIterator.moveNext (dart:core/runtime/libcore_patch.dart:151:12)
flutter: #12     Iterable.single (dart:core/iterable.dart:552:13)
flutter: #13     RenderObject._updateSemantics (package:flutter/src/rendering/object.dart:2327:7)
flutter: #14     PipelineOwner.flushSemantics (package:flutter/src/rendering/object.dart:880:16)
flutter: #15     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:274:19)
flutter: #16     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:654:13)
flutter: #17     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:208:5)
flutter: #18     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990:15)
flutter: #19     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:930:9)
flutter: #20     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:842:5)
flutter: #21     _invoke (dart:ui/hooks.dart:128:13)
flutter: #22     _drawFrame (dart:ui/hooks.dart:117:3)
flutter: (elided 2 frames from class _AssertionError)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

**Comment by code1up** [comment]
Flutter `doctor` output:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v0.7.3, on Mac OS X 10.13.6 17G65, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
[✓] Android Studio (version 3.0)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.27.1)
[✓] Connected devices (1 available)
```

**Comment by code1up** [comment]
Tested on iOS Simulator only.

**Comment by zoechi** [comment]
I was not able to reproduce
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, v0.7.5, on Mac OS X 10.13.6 17G65, locale en-AT)
[✓] Android toolchain - develop for Android devices (Android SDK 28.0.1)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
[✓] Android Studio (version 3.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2018.2)
[✓] Connected devices (1 available)

• No issues found!
```

**Comment by code1up** [comment]
Gunter

Thanks for trying and the quick response - appreciated. Repro here is simply to rapidly scroll up and down. Seeing those exceptions in the VS Code logs.

I created this standalone example from a much larger project where this issue is occurring.

Looks like I'm on an older version too. Will upgrade and retry.

**Comment by zoechi** [comment]
Yes, I'd suggest to try with `master` channel.
I tried again scrolling rapidly, but I don't see any error output.

**Comment by code1up** [comment]
Reproduced:

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, v0.8.1, on Mac OS X 10.13.6 17G65, locale en-GB)
[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
[✓] Android Studio (version 3.0)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.27.1)
[✓] Connected devices (1 available)
```

**Comment by zoechi** [comment]
> Or can these exceptions be ignored

They shouldn't be ignored. The app probably won't behave properly after such an exception.

Can you please post the `flutter doctor -v` output instead of just `flutter doctor`?

Have you tried on different devices or emulator?
What platform iOS/Android?

**Comment by code1up** [comment]
[✓] Flutter (Channel dev, v0.8.1, on Mac OS X 10.13.6 17G65, locale en-GB)
    • Flutter version 0.8.1 at /Users/alan.gorton/p/ag/frameworks/flutter
    • Framework revision 0235ec59d1 (2 days ago), 2018-09-06 20:23:13 -0700
    • Engine revision 9b297b3b51
    • Dart version 2.1.0-dev.3.1.flutter-760a9690c2

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/alan.gorton/Library/Android/sdk
    • Android NDK at /Users/alan.gorton/Library/Android/sdk/ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] VS Code (version 1.27.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.18.0

[✓] Connected devices (1 available)
    • iPhone X • FFB505A7-CBC9-4034-91C1-C42F2DC225F1 • ios • iOS 11.4 (simulator)

• No issues found!

**Comment by code1up** [comment]
Günter

Again - many thanks.

So just tried on iPhone X Simulator. Can repro from VS Code / `flutter run` command line.

Will try on real device ASAP.

We're in violent agreement re not ignoring exceptions!

**Comment by code1up** [comment]
Tested on iPhone 7 using `flutter run`.

Seeing another exception (change list size to 1K [not a realistic production value for keep alive, just to stress]):

`flutter: Another exception was thrown: NoSuchMethodError: The method 'visitChildren' was called on null.`

**Comment by zoechi** [comment]
Is the original issue solved?

It's not a good idea to reuse an issue for a different problem.

I think

> The method 'visitChildren' was called on null.

should be a new issue.

**Comment by alangorton** [comment]
The original issue is not solved. When progress is made on this ticket, the `visitChildren` issue might be solved too. Reluctant to raise a second issue when the original issue is unresolved.

Not sure how to find out if the is actively being considered as a bug and potentially worked on. The repro is straightforward and self-contained.

Thanks.

**Comment by mehmetf** [comment]
I so wanted to reproduce this because we are seeing a similar problem as well but only from certain configurations and versions of the app.

However, I could not. The original repro on an iOS simulator works for me without any errors :-/ I even increased the widget count to 500 and used this crazy widget as a loading indicator:

```dart
     return Padding(
        padding: const EdgeInsets.symmetric(vertical: 8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          children: [
            Icon(Icons.warning),
            Padding(padding: EdgeInsets.symmetric(horizontal: 10.0)),
            Text(
              'WAITING',
              style: theme.textTheme.display1,
            )
          ],
        ),
      );
```

[✓] Flutter (Channel master, v0.8.6-pre.511, on Mac OS X 10.13.6 17G3025, locale en-US)
    • Flutter version 0.8.6-pre.511 at /Users/mehmetf/flutter/mehmetf/flutter
    • Framework revision 57c2fac14a (5 days ago), 2018-11-07 13:06:38 -0800
    • Engine revision 2494abdda7
    • Dart version 2.1.0 (build 2.1.0-dev.8.0 bf26f760b1)

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.2)
    • Android SDK at /Users/mehmetf/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-stable, build-tools 28.0.2
    • Java binary at: /Applications/Android Studio 3.3 Preview.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] IntelliJ IDEA Community Edition (version 2018.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 27.1.2
    • Dart plugin version 181.4203.498

[✓] Connected device (1 available)
    • iPhone X • 2F9A5867-7FC8-43C0-BE5E-383782EECA81 • ios • iOS 11.2 (simulator)

**Comment by alangorton** [comment]
@mehmetf thanks for the update. I need to take another look at the repro with the latest and greatest Flutter. Will report back. Cheers.

**Comment by goderbauer** [comment]
@alangorton were you able to confirm that your issue is fixed now? We recently fixed a few issues in this area.

**Comment by zoechi** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. We are therefore reluctantly going to close this bug for now. Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!
Thanks for your contribution.

**Comment by alangorton** [comment]
Can confirm this appears to be fixed in 1.0 stable. Thanks.
